pkg/util/metrics/problem_gauge.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package metrics
+
+import (
+	"context"
+	"fmt"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+type ConditionGauge struct {
+	metricName string
+	keyReason  tag.Key
+	reason     string
+	gauge      *stats.Int64Measure
+}
+
+func NewConditionGauge(metricName string, description string) (ConditionGauge, error) {
+	cg := ConditionGauge{metricName: metricName}
+
+	var err error
+	cg.keyReason, err = tag.NewKey("reason")
+	if err != nil {
+		return cg, err
+	}
+	cg.gauge = NewInt64Metric(
+		cg.metricName,
+		description,
+		"",
+		view.LastValue(),
+		[]tag.Key{cg.keyReason})
+	return cg, nil
+}
+
+func (cg ConditionGauge) Set(reason string) error {
+	if cg.gauge == nil {
+		return fmt.Errorf("condition gauge %q is being set before initialized.", cg.metricName)
+	}
+
+	// If the condition reason has changed, we should clear the previous condition reason.
+	if cg.reason != "" && cg.reason != reason {
+		if err := cg.Clear(); err != nil {
```

Comment:
Why do we need to clear the previous condition reason?

Reply:
Here is the tricky thing. A metrics with different labels are actually considered as two independent metrics.

Let's first see how node condition behaves:

1. At start, it looks like this:
```
{  
   "type":"KernelDeadlock",
   "status":"False",
   "transition":"...",
   "reason":"KernelHasNoDeadlock",
   "message":"kernel has no deadlock"
}
```
2. When a problem happens, it becomes this:
```
{  
   "type":"KernelDeadlock",
   "status":"True",
   "transition":"...",
   "reason":"DockerHung",
   "message":"task dockerd blocked for..."
}
```
3. If a new type of kernel deadlock happens, after condition manager reports to apiserver, it becomes this:
```
{  
   "type":"KernelDeadlock",
   "status":"True",
   "transition":"...",
   "reason":"ContainerdHung",
   "message":"task containerd blocked for..."
}
```
4. If someone then clears the condition, it will become this:
```
{  
   "type":"KernelDeadlock",
   "status":"False",
   "transition":"...",
   "reason":"KernelHasNoDeadlock",
   "message":"kernel has no deadlock"
}
```

However, for metrics, given the fact that below three metrics are independent, it becomes tricky:

- kernel_task_hung{reason=""} 0
- kernel_task_hung{reason="DockerHung"} 1
- kernel_task_hung{reason="ContainerdHung"} 1

Now, let's do the same workflow again, assuming that we do not clear the previous `reason`:
1. At start, it looks like this:
```
- kernel_task_hung{reason=""} 0
```
2. When a problem happens, it becomes this (two metrics):
```
- kernel_task_hung{reason=""} 0
- kernel_task_hung{reason="DockerHung"} 1
```

3. If a new type of kernel deadlock happens, it becomes this (three metrics):
```
- kernel_task_hung{reason=""} 0
- kernel_task_hung{reason="DockerHung"} 1
- kernel_task_hung{reason="ContainerdHung"} 1
```
4. If someone then clears the condition, it will become this:
```
- kernel_task_hung{reason=""} 0
- kernel_task_hung{reason="DockerHung"} 0
- kernel_task_hung{reason="ContainerdHung"} 0
```

Now let's do the same workflow again, assuming that we clear the previous `reason` (current PR implementation):
1. At start, it looks like this:
```
- kernel_task_hung{reason=""} 0
```
2. When a problem happens, it becomes this (two metrics):
```
- kernel_task_hung{reason=""} 0
- kernel_task_hung{reason="DockerHung"} 1
```

3. If a new type of kernel deadlock happens, it becomes this (three metrics):
```
- kernel_task_hung{reason=""} 0
- kernel_task_hung{reason="DockerHung"} 0
- kernel_task_hung{reason="ContainerdHung"} 1
```
4. If someone then clears the condition, it will become this:
```
- kernel_task_hung{reason=""} 0
- kernel_task_hung{reason="DockerHung"} 0
- kernel_task_hung{reason="ContainerdHung"} 0
```

Note how step 3 in above two cases are different, depending on whether we clear the previous reason or not:

- By clearing the previous `reason`, it is more similar to what we have with node condition: at any given point of time, only the most recent `reason` can be true (non-zero);
- By not clearing the previous `reason`, we obtain information about what are the `reason` that has happened over the history.

I chose to clear the `reason` simply to stay closer with node condition's implementation. I'm happy if we choose to not to clear the `reason` as well.

Do you have any preference on which one to choose from the above two options?

Reply:
I see your points now. Clearing the previous reason is fine. But when the condition is false, I think the reason should not be "", but should be something like "KernelHasNoDeadlock".

Reply:
>  But when the condition is false, I think the reason should not be "", but should be something like "KernelHasNoDeadlock".

I agree. Sorry this is a bug in my first implementation.

This bug happened, because NPD set condition "KernelDeadlock" to "Status: false, Reason:KernelHasNoDeadlock" upon NPD start. And in my initial implementation, I'm setting the metrics based on it.

In the current implementation, at file `pkg/systemlogmonitor/log_monitor.go` function `initialProblemGauges`, there is this piece of code:
```
for _, condition := range initialConditions {
if condition.Status != types.True {
	continue
}
```

That way I can avoid setting a `problem_gauge` metrics with these nonsense reasons (e.g. "KernelHasNoDeadlock").

I also updated the README and PR description to reflect this change.